### ISSUE TITLE
Keyboard accessibility for input controls

### ISSUE THREAD

**Issue by OliverJAsh** [google/material-design-lite#805]
http://www.getmdl.io/components/index.html#toggles-section

I can't switch between radio button options using my keyboard, unlike native (non-enhanced) radio buttons.

**Comment by sgomes** [comment]
Thanks, @OliverJAsh, we must have had a regression here, as I remember this working earlier. I'll investigate.

**Comment by HenriqueLimas** [comment]
@sgomes , it seems that we are removing the focus event that input has when you click in it. I see that we have a hack at [radio.js](https://github.com/google/material-design-lite/blob/master/src/radio/radio.js#L149) that "remove" input focus. We also have that in others toggles input like [checkbox](https://github.com/google/material-design-lite/blob/master/src/checkbox/checkbox.js#L138) and [switch](https://github.com/google/material-design-lite/blob/a3bfcf45209d1f5356fd3288cb1b1f83a1923dc2/src/switch/switch.js#L137).

**Comment by OliverJAsh** [comment]
Also appears these elements don't have focus states. Should they?

**Comment by leifoolsen** [comment]
Keyboard handling for MDL toggles does not behave as standard radiobuttons and checkboxes: TAB-key should move to next tabstop. SPACE-key should toggle on/off.

**Comment by sgomes** [comment]
Thanks, @leifoolsen. Is the the case for all toggles, or just radio buttons, as the issue mentions? Because I believe in the specific case of radio buttons, it's actually Up and Down, which should work here.

**Comment by leifoolsen** [comment]
@sgomes: Here is HTML for my test case:

<h4>Toggles</h4>
<div class="mdl-grid">
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-checkbox mdl-js-checkbox mdl-js-ripple-effect" for="checkbox-1">
      <input type="checkbox" id="checkbox-1" class="mdl-checkbox__input" checked>
      <span class="mdl-checkbox__label">Checkbox</span>
    </label>
    <p>Check on</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-checkbox mdl-js-checkbox mdl-js-ripple-effect" for="checkbox-2">
      <input type="checkbox" id="checkbox-2" class="mdl-checkbox__input">
      <span class="mdl-checkbox__label">Checkbox</span>
    </label>
    <p>Check off</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-checkbox mdl-js-checkbox mdl-js-ripple-effect" for="checkbox-3">
      <input type="checkbox" id="checkbox-3" class="mdl-checkbox__input" checked disabled>
      <span class="mdl-checkbox__label">Checkbox</span>
    </label>
    <p>Check disabled</p>
  </div>

  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-radio mdl-js-radio mdl-js-ripple-effect" for="option-1">
      <input type="radio" id="option-1" class="mdl-radio__button" name="options" value="1" checked>
      <span class="mdl-radio__label">First</span>
    </label>
    <p>Radio on</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-radio mdl-js-radio mdl-js-ripple-effect" for="option-2">
      <input type="radio" id="option-2" class="mdl-radio__button" name="options" value="2">
      <span class="mdl-radio__label">Second</span>
    </label>
    <p>Radio off</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-radio mdl-js-radio mdl-js-ripple-effect" for="option-3">
      <input type="radio" id="option-3" class="mdl-radio__button" name="options" value="3" disabled>
      <span class="mdl-radio__label">Third</span>
    </label>
    <p>Radio disabled</p>
  </div>

  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-icon-toggle mdl-js-icon-toggle mdl-js-ripple-effect" for="icon-toggle-1">
      <input type="checkbox" id="icon-toggle-1" class="mdl-icon-toggle__input" checked>
      <i class="mdl-icon-toggle__label material-icons">format_bold</i>
    </label>
    <p>Icon on</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-icon-toggle mdl-js-icon-toggle mdl-js-ripple-effect" for="icon-toggle-2">
      <input type="checkbox" id="icon-toggle-2" class="mdl-icon-toggle__input">
      <i class="mdl-icon-toggle__label material-icons">format_italic</i>
    </label>
    <p>Icon off</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-icon-toggle mdl-js-icon-toggle mdl-js-ripple-effect" for="icon-toggle-3">
      <input type="checkbox" id="icon-toggle-3" class="mdl-icon-toggle__input" disabled>
      <i class="mdl-icon-toggle__label material-icons">format_italic</i>
    </label>
    <p>Icon disabled</p>
  </div>

  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-switch mdl-js-switch mdl-js-ripple-effect" for="switch-1">
      <input type="checkbox" id="switch-1" class="mdl-switch__input" checked>
      <span class="mdl-switch__label"></span>
    </label>
    <p>Switch on</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-switch mdl-js-switch mdl-js-ripple-effect" for="switch-2">
      <input type="checkbox" id="switch-2" class="mdl-switch__input">
      <span class="mdl-switch__label"></span>
    </label>
    <p>Switch off</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label class="mdl-switch mdl-js-switch mdl-js-ripple-effect" for="switch-3">
      <input type="checkbox" id="switch-3" class="mdl-switch__input" disabled>
      <span class="mdl-switch__label"></span>
    </label>
    <p>Switch disabled</p>
  </div>
</div>

<h4>Native Toggles</h4>
<div class="mdl-grid">
  <div class="mdl-cell mdl-cell--2-col">
    <label for="native-checkbox-1">
      <input type="checkbox" id="native-checkbox-1" checked>
      <span>Checkbox</span>
    </label>
    <p>Check on</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label for="native-checkbox-2">
      <input type="checkbox" id="native-checkbox-2">
      <span>Checkbox</span>
    </label>
    <p>Check off</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label for="native-checkbox-3">
      <input type="checkbox" id="native-checkbox-3" checked disabled>
      <span>Checkbox</span>
    </label>
    <p>Check disabled</p>
  </div>

  <div class="mdl-cell mdl-cell--2-col">
    <label for="native-option-1">
      <input type="radio" id="native-option-1" name="native-options" value="1" checked>
      <span>First</span>
    </label>
    <p>Radio on</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label for="native-option-2">
      <input type="radio" id="native-option-2" name="native-options" value="2">
      <span>Second</span>
    </label>
    <p>Radio off</p>
  </div>
  <div class="mdl-cell mdl-cell--2-col">
    <label for="native-option-3">
      <input type="radio" id="native-option-3" name="native-options" value="3" disabled>
      <span class="mdl-radio__label">Third</span>
    </label>
    <p>Radio disabled</p>
  </div>
</div>

<h4>Sliders</h4>
<div class="mdl-grid">
  <div class="mdl-cell mdl-cell--4-col mdl-cell--2-col-phone">
    
    <input class="mdl-slider mdl-js-slider" type="range" min="0" max="100" value="0">

    <p>Default slider	</p>
  </div>

  <div class="mdl-cell mdl-cell--4-col mdl-cell--2-col-phone">
    
    <input class="mdl-slider mdl-js-slider" type="range" min="0" max="100" value="25">

    <p>Starting value</p>
  </div>
</div>


1: If I tab into the first checkbox, then I can use SPACE to toggle.
2: If I click on the checkbox, then I can NOT use SPACE to toggle.
3: If I have clicked on a checkbox and then press TAB, focus moves to my header section. I do not have any tabindex set in my header.
4: Radiobuttons never get focus when TAB'ing.
5: Native toggles works as expected: TAB-key focuses the checkboxes and the selected radiobutton, arrow keys moves focus to next radiobutton. The "name" attribute should be sufficient for navigating trough grouped radiobuttons.
6: In general, clicking on one of the MDL toggles before using the TAB-key messes up the "natural" tab order.
I get roughly the same behavior on the MDL component page: http://www.getmdl.io/components/index.html#toggles-section using Chrome on OSX.

**Comment by sgomes** [comment]
Thank you, @leifoolsen, that's a very thorough and detailed test case! Looks like we have a bunch to fix.

We'll be making changes to how all those components are implemented in 2.0, which will hopefully address these issues. Since we'll no longer need to support old browsers with fallback native elements, we can explore different approaches like doing everything in ARIA, which will hopefully lead to fewer bugs.

Would you mind if I poked you when we have something ready in a few months, so you can take a look and do a little sanity check for us? :)

**Comment by leifoolsen** [comment]
@sgomes: Looking forward to version 2.0. Please notify me when the new version is ready, I'll be more than happy to contribute with some sanity checking. BTW: I have a GIT project showcasing all components in MDL version 1.1.1. You can check it out here if you like: https://github.com/leifoolsen/mdl-webpack

**Comment by sgomes** [comment]
Fantastic, thank you, @leifoolsen!

The showcase looks great! By the way, we've made a few changes to our samples after 1.1.0 (particularly around lists), so you might want to pull those in.

**Comment by leifoolsen** [comment]
Thank you @sgomes! Updated my lists markup to 1.1.1. Looks much better in than in 1.1.0.

**Comment by tschiela** [comment]
Is this still open? Seems to be fixed

**Comment by leifoolsen** [comment]
@tschiela 
Don't think this is fixed. See my test case above.

@sgomes 
Regarding keyboard handling: I think you have to completely  rewrite the checkbox and radiobutton components. There is no way to keep a natural tab order with today's generated markup, e.g.

```html
<label class="mdl-checkbox mdl-js-checkbox mdl-js-ripple-effect" for="checkbox-1">
  <input type="checkbox" id="checkbox-1" class="mdl-checkbox__input" checked>
  <span class="mdl-checkbox__label">Checkbox</span>
</label>
```
becomes:
```html
<label class="mdl-checkbox mdl-js-checkbox mdl-js-ripple-effect mdl-js-ripple-effect--ignore-events is-upgraded is-checked" for="checkbox-1" data-upgraded=",MaterialCheckbox,MaterialRipple">
  <input type="checkbox" id="checkbox-1" class="mdl-checkbox__input" checked="">
  <span class="mdl-checkbox__label">Checkbox</span>
  <span class="mdl-checkbox__focus-helper"></span>
  <span class="mdl-checkbox__box-outline">
    <span class="mdl-checkbox__tick-outline"></span>
  </span>
  <span class="mdl-checkbox__ripple-container mdl-js-ripple-effect mdl-ripple--center" data-upgraded=",MaterialRipple">
    <span class="mdl-ripple"></span>
    </span>
</label>
```
The only element in the generated markup which can receive focus is the input element - and that element is hidden.

**Comment by tschiela** [comment]
@leifoolsen In the showcase both form elements are focusable. Checkbox can be selected with spacebar and is focusable. Also radio buttons are focusable and can be changed by arrow keys.  Just the radios need some focus-styling. 

To hide the input elements i would suggest some screen reader only css rules.  For example from H5BP:
.sr-only {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

The 0 dimension will not work with the screenreader on Mac OS X Safari (Yosemite). 

Also it seems to be to mutch JavaScript here for some simple stuff like styling (except the ripple effect) radio btns and checkboxes. The whole styling and focusing can be achieved with plain html and css.

**Comment by Garbee** [comment]
Yes, we fixed the radio demo. The problem was caused due to a new component that copied the example code in, which created duplicate ID entries in the component page. That then of course messed up engines referencing the right pieces since IDs shouldn't be duplicated.

As far as natural tab ordering, V2 is entirely re-wiring the internals to components and providing developers full access to what is happening. When alpha is announced we'd love to get input as we develop from then on with the a11y side of things to make sure we are making things that are as good as they can be.

Since the primary issue report here is fixed, I'm closing this out since there isn't anything directly actionable to carry out.

Thank you.

**Comment by leifoolsen** [comment]
@tschiela Yes, the checkboxes are focusable and you can use space for toggling. The problem occurs if you click one of the check boxes and then use the TAB key to move to the next focusable item - the focus i lost. I also agree that the CSS is to complicated. Some of the generated code could be replaced with :before and :after pseudo classes I believe.
